### PR TITLE
Allow to diff two configurations and return the difference in the keys.

### PR DESCRIPTION
**Pull request by ph**
Pass 2 difference Config object to DiffKeys and it will return a Diff
type that will contains the following:

- Added keys
- Keep keys
- Removed keys

Add a few helpers method on the type to know if something is different
between the 2 configuration.

### CODE REVIEW

**Comment by ph** [comment]
@urso When working on the implicit keystore usage in beats, I needed something like this.

File: diff.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+				p := ctx.path(opts.pathSep)
+				keys = append(keys, p)
+			} else {
+				newKeys := FlattenKeys(opts, scfg)
+				keys = append(keys, newKeys...)
+			}
+		}
+	} else {
+		// We should not get here
+		panic("unknown config type")
```
Comment (ph) [inline diff.go]: I believe this should not happen but I think it could.

**Comment by urso** [comment]
+1

In go-ucfg, the top-level package is mostly the core for packing/unpacking/querying the config object. All extra functionality is in sub-packages.

As getters/setters can operate on complete settings paths, adding the `FlattenKeys` as method to `*Config` does totally make sense to me.

**Comment by ph** [comment]
@urso I have applied the changes we were talking about.

File: diff/keys.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
+package diff
+
+import (
+	"fmt"
+	"strings"
+
+	ucfg "github.com/elastic/go-ucfg"
```
Comment (urso) [inline diff/keys.go]: Note: This line is equivalent to `"github.com/elastic/go-ucfg"`.

goimports normaly adds this kind of line. If `-` is used in package names, the later part of the name makes the package.

File: diff/keys.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	return false
+}
+
+// GoStringer implement the GoStringer interface
+func (d Diff) GoStringer() string {
+	return d.String()
+}
+
+// Keys takes two configuration and return the difference between the defined keys
+func Keys(old, new *ucfg.Config, opts ...ucfg.Option) Diff {
```
Comment (urso) [inline diff/keys.go]: This is basically the constructor. How about naming it one of `(Make/Build/Create)Diff`, `Compare`, `CompareConfigs` or `Changes`?

File: ucfg.go [hunk 4]
```diff
@@ -132,6 +133,46 @@ func (c *Config) Parent() *Config {
 	}
 }
 
+// FlattenedKeys return a flattened views of the set keys in the configuration
+func (c *Config) FlattenedKeys(opts ...Option) []string {
+	var keys []string
+	normalizedOptions := makeOptions(opts)
```
Comment (urso) [inline ucfg.go]: Not sure, but `pathSep` might be empty here. If so add:

```
if normalizedOptions.pathSep == "" {
  normalizedOptions.pathSep = "."
}
```

File: ucfg.go [hunk 5]
```diff
@@ -132,6 +133,46 @@ func (c *Config) Parent() *Config {
...
+				p := ctx.path(normalizedOptions.pathSep)
+				keys = append(keys, p)
+			} else {
+				newKeys := scfg.FlattenedKeys(opts...)
+				keys = append(keys, newKeys...)
+			}
+		}
+	} else {
+		// We should not get here
+		panic("unknown config type")
```
Comment (urso) [inline ucfg.go]: hm... not sure about 'empty' config objects.

File: ucfg.go [hunk 6]
```diff
@@ -132,6 +133,46 @@ func (c *Config) Parent() *Config {
...
+	}
+
+	sort.Strings(keys)
```
Comment (urso) [inline ucfg.go]: We need to sort here? If so the godoc should reflect the list returned is always being sorted.

**Comment by ph** [comment]
@urso I choose `CompareConfigs`, I think it will be more explicit that way.

File: ucfg_test.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		"values.1.j",
+		"values.1.o",
+	}
+	sort.Strings(expected)
+
+	results := c.FlattenedKeys(opts...)
+	sort.Strings(results)
+
+	assert.Equal(t, expected, results)
+}
```
Comment (urso) [inline ucfg_test.go]: tests are quite similar. Rather introduce a table-driven test + use `Run` like:

```
func TestFlattenKeys(t *testing.T) {
  tests := struct {
    name string
    pathSep string
  }{
    { "withDot", "." },
    { "emptySep", "" },
  }

  sorted := func(s []string) []string {
    sort.Strings(s)
    return s
  }

  cfg := map[string]interface{}{ ... }
  expected := sorted([]string { ... })

  for _, test := range tests {
    sep := test.pathSep
    t.Run(test.name, func(t *testing.T) {
      opts := []Option{}
      if sep != "" {
        opts = append(opts, PathSep(sep))
      }

      c, err := NewFrom(cfg, opts...)
      if err != nil { t.Fatal(err) }
      assert.Equal(t, expected, sorted(c.FlattenKeys(opts...)))
    })
  }
}
```

**Comment by ph** [comment]
@urso, I've moved the two similar test into a table driven test.

**Comment by ph** [comment]
seems to be a problem with the ci, I see that you are working on it on #95.

**Comment by ph** [comment]
rebased with master, the ci should run.

**Pull request merged**